_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,275 @@
+meta_description: Learn about different searches in OpenSearch. Discover how hybrid search merges results from multiple query sources for improved relevance.
+---
+
+In the evolving world of search technology, OpenSearch stands out with its powerful search capabilities. OpenSearch offers a range of search techniques, starting with basic lexical search using BM25 (Best Matching 25), which excels at keyword-based retrieval and remains a cornerstone of many search applications. The evolution continues with semantic search, which leverages dense vector embeddings to capture the semantic meaning beyond keywords. Advanced hybrid techniques combine lexical and semantic approaches, using normalization methods like score-based and rank-based algorithms to balance and integrate scores from different search methods. These techniques progressively enhance search relevance and flexibility, catering to diverse search requirements in modern applications. This blog post will guide you through this series of search techniques, from BM25 to hybrid search and provide an overview of how Hybrid Search works in OpenSearch.
+
+
+### 1. Lexical Search vs Semantic Search
+
+#### Lexical Search (BM25) 
+
+BM25 (Best Matching 25) is a probabilistic retrieval framework used as the default scoring algorithm in many search engines, including OpenSearch. In OpenSearch, BM25 is implemented as part of the standard full-text search capabilities. It can be easily applied across multiple fields using the multi match, range or term queries allowing for field-specific boosting and flexible matching criteria. This makes it a versatile choice for a wide range of search applications, from e-commerce product searches to document retrieval systems.
+
+API Request:
+
+
+```
+GET /bookstore_catalog/_search
+{
+  "query": {
+    "match": {
+      "title": "the martian"
+    }
+  }
+}
+```
+
+
+
+#### Semantic Search
+
+
+Semantic search in OpenSearch represents a significant advancement in search technology, leveraging the power of machine learning and natural language processing. This approach uses dense vector embeddings to represent both documents and queries in a high-dimensional space, allowing for semantic similarity comparisons that go beyond simple keyword matching. At the core of semantic search is the concept of text embeddings - dense vector representations of text that capture semantic meaning. These embeddings are typically generated using large language models trained on vast corpora of text. In OpenSearch, these embeddings are stored as dense vector fields and can be efficiently searched using k nearest neighbor algorithms. At search time, the user's query is similarly transformed into a vector embedding. The system then finds documents whose vector representations are closest to the query vector, typically using cosine similarity or Euclidean distance.
```

Comment:
I'm not sure that is correct statement: "typically using cosine similarity or Euclidean distance." Exact space type is use-case dependent, and I see inner product/dot product far more often. You can frame it something like "The system then finds documents whose vector representations are closest to the query vector, using different space types like inner product, cosine similarity, and others to measure the similarity between the query and document embeddings."